Toggle JSON unfolding by field

calherries opened the pull request:
This PR resolves https://github.com/metabase/metabase/issues/24801 and milestone 1 of https://github.com/metabase/metabase/issues/28667.

I created a new column in the `metabase_field` table: `json_unfolding`. The column is false by default for non-JSON columns, and for existing JSON fields, it should take the value of `(get-in database [:details :json-unfolding])` or true if it's not set. This means json_unfolding will be true for a field if the database type of the column is 'type/JSON' and `nested-field-columns` is not supported by a driver. This is by design, so that JSON fields will be automatically unfolded if `nested-field-columns` is eventually supported by a driver, without the need for a migration on the `json_unfolding` column. None of the 3rd party drivers [listed in the Metabase docs](https://www.metabase.com/docs/latest/developers-guide/partner-and-community-drivers) support `nested-field-columns` anyway though.

To change json_unfolding for a field, I've created a new field setting that is visible if the base_type of a field is type/JSON and the driver supports `nested-field-columns`:
![image](https://user-images.githubusercontent.com/39073188/223469986-58c1e563-3248-40d3-9af3-d02fb68a5b8c.png).

If "Yes", then the JSON field will be unfolded on the next sync. If "No", then any unfolded JSON fields from that field will be inactivated on the next sync. The next milestone will make the sync automatically apply when the setting changes.

You can test this feature out with the json test dataset at `test/metabase/test/data/dataset_definitions/json.edn`. To create the DB run a test that uses it like `json-unfolding-default-true-test` with postgres or mysql.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/28742)
<!-- Reviewable:end -->

Commit:
Migration for adding nfc_enabled

Commit:
Add nfc_enabled to field settings

Commit:
Update describe-nested-field-columns to only unfold fields that have not been disabled

Commit:
Remove spy

Commit:
Tidy

Commit:
Fix

Commit:
Clear nested fields immediately if folding is disabled

Commit:
Clear nested fields on nfc_enabled change

Commit:
Trim trailing whitespace

Commit:
Tidy

Commit:
Fix

Commit:
Add enable-json-unfolding-test

Commit:
Move to field api test

Commit:
nfc_enabled -> json_unfolding

Commit:
Tidy test

Commit:
Make json-unfolding in database details just the default for new settings

Commit:
Restore original visibility-type logic

Commit:
Fix

Commit:
Sync field json_unfolding according to db json_folding

Commit:
Rename to json-unfolding-default

Commit:
Add test for the case when the json-unfolding is false for the database

Commit:
Implement default json unfolding for first sync

Commit:
Update comment

Commit:
Update comment

Commit:
Restore PUT field

Commit:
Migration for populating json_unfolding for mysql and postgres

Commit:
Remove migration and use default on the frontend instead

Commit:
Update json-unfolding database setting copy

Commit:
Move Unfold JSON setting under semantic type setting

calherries commented on frontend/src/metabase/admin/datamodel/containers/FieldApp.jsx:
```diff
@@ -326,6 +355,7 @@ const FieldGeneralPane = ({
         />
       </Section>
     )}
+
```
The other sections have empty lines separating them, so I added one here

Commit:
Separate sentences with spaces

Commit:
Use type/JSON instead of type/SerializedJSON

Commit:
Capitalize

Commit:
Tidy migration

Commit:
Update migration

Commit:
Fix rollback migration for h2

Commit:
whitespace

Commit:
Add test for migration

Commit:
Fix test

Commit:
Add rollback

Commit:
whitespace

Commit:
Test JSONB type as well as JSON

Commit:
Don't fingerprint JSON columns

Commit:
Remove comment, that might be wrong in the future

Commit:
Update test

Commit:
Add tests for fingerprinting

Commit:
Restore handleChangeSemanticType

calherries commented on frontend/src/metabase/admin/datamodel/containers/FieldApp.jsx:
```diff
@@ -297,6 +300,32 @@ const FieldGeneralPane = ({
       />
     </Section>
 
+    {isa(field.semantic_type, TYPE.SerializedJSON) &&
```
This isn't great because a user can change the semantic_type of a field. But we can only unfold JSON for fields with the correct database types.

https://github.com/metabase/metabase/pull/28849 should fix it.

Commit:
Use base_type TYPE.JSON instead

calherries commented on frontend/src/metabase/admin/datamodel/containers/FieldApp.jsx:
```diff
@@ -297,6 +300,32 @@ const FieldGeneralPane = ({
       />
     </Section>
 
+    {isa(field.base_type, TYPE.JSON) &&
```
This depends on https://github.com/metabase/metabase/pull/28849

Commit:
Use base-type JSON for fingerprinting base query

Commit:
Add test for visibility-type=details-only

Commit:
undo .

calherries commented on test/metabase/api/field_test.clj:
```diff
@@ -724,3 +725,56 @@
                                     [3 "Synergistic Granite Chair"]]}
                           (mt/user-http-request :crowberto :get 200 (format "field/%d/values" (mt/id :orders :product_id)))))))))))
+
+(deftest json-unfolding-default-true-test
```
I created two tests here, one for when `json-unfolding=true`, and another when `json-unfolding=false`

calherries commented on src/metabase/driver/mysql.clj:
```diff
@@ -48,11 +48,7 @@
-    (if (nil? json-setting)
-      true
-      json-setting)))
+(defmethod driver/database-supports? [:mysql :nested-field-columns] [_driver _feat _db] true)
```
This is part of a refactor. Now, `database-supports?` just returns a boolean based on the driver. The `database.details.json-unfolding` setting is checked by the caller of this function separately if needed, e.g. here https://github.com/metabase/metabase/pull/28742/files#diff-e4db3baaafcba181fc735b94c34048deb437fc47e8a2bd3acb9057732d92b330R159-R161

qnkhuat replied:
are we sure this will be true by default? json only supports from MySQL 5.7.8, I think we want to support 5.7?

calherries replied:
If that's the case, why do the JSON tests pass for MySQL? We're using 5.7.23 in CI

calherries replied:
I checked our docs, and you're right. We're meant to support MySQL 5.7.7, which doesn't have the JSON type. It's a pity we're not testing against this in CI. But since JSON unfolding will only apply to JSON types, it won't cause any bugs. It's the same scenario as redshift was in until today, with [this PR](https://github.com/metabase/metabase/pull/29465). The only bug visible to users (which currently exists) would be that the "Unfold JSON by Default" setting would be present in the MySQL DB's connection settings page, regardless of its version.

calherries replied:
I added a comment explaining this https://github.com/metabase/metabase/commit/e83d67498f699b504b0a8e93ea46fd65033fc24e

qnkhuat replied:
FWIW

5.7 will EOL in 10 months.

https://endoflife.date/mysql

Commit:
Change order of migrations

Commit:
Change migration id

calherries commented on resources/migrations/000_migrations.yaml:
```diff
@@ -14354,8 +14354,45 @@ databaseChangeLog:
+                  nullable: true
+            tableName: metabase_field
+
+  - changeSet:
```
To preserve unfolding of existing fields, this backfills the new `metabase._field.json_unfolding` column to equal true if `metabase_database.details.json-unfolding` is true or nil.

calherries committed:
Merge branch 'master' into type-json-base-type-postgres-mysql

Commit:
Fix migration test

Commit:
Add h2 migration

calherries committed:
Update test description

Commit:
Add test for migration

calherries committed:
Remove validCheckSum

Co-authored-by: Ngoc Khuat <qn.khuat@gmail.com>

qnkhuat commented:
@calherries I'm testing this and I think I found 2 bugs, recorded a loom [here](https://www.loom.com/share/bfe92a49ddeb412983e3ddd203dac98b)

Commit:
Use ?? instead of ||

Commit:
Fix tests

Commit:
Merge master

Commit:
Merge branch 'master' into type-json-base-type-postgres-mysql

Commit:
Exclude mariadb from tests

Commit:
Make is-mariadb? public

Commit:
Remove outdated serdes stuff

Commit:
Remove unnecessary and

Commit:
Fix mysql migration

calherries commented:
@qnkhuat The first bug in your loom (details-only visibility not working) is a regression on master, which I'll file in a sec. I can't reproduce the second bug from 1:10 in the recording (the something went wrong bug). There's a good chance I've fixed it though. Can you try again?

Commit:
Remove unused require

qnkhuat commented on resources/migrations/000_migrations.yaml:
```diff
@@ -14362,6 +14361,62 @@ databaseChangeLog:
+                name: json_unfolding
+                type: boolean
+                constraints:
+                  nullable: true
```
should we have another migration to make this non-nullable and have a default value?

FWIW I asked QP ppl about a similar problem for `database_requried` [here](https://metaboat.slack.com/archives/C04DN5VRQM6/p1678112076860729).

And the consensus is that we should have default value and make it nullable.

calherries replied:
True, json_unfolding could be `false` by default. Could prevent some footguns. Good idea

qnkhuat commented on resources/migrations/000_migrations.yaml:
```diff
@@ -14362,6 +14361,62 @@ databaseChangeLog:
+      id: v47.00-003
+      author: calherries
+      comment: Added 0.47.0 - Populate metabase_field.json_unfolding based on database details
+      validCheckSum: ANY
```
```suggestion
```

Commit:
Fix test

qnkhuat commented on test/metabase/db/schema_migrations_test.clj:
```diff
@@ -1000,3 +1000,49 @@
+             enabled-by-default-db-2-id
+             disabled-db-id] (t2/insert-returning-pks! Database [{:name    "enabled"
+                                                                  :engine  "postgres"
+                                                                  :details {:json-unfolding true}}
```
worth adding a case where `json-unfolding` key is not present? Not sure if it could lack that key in the wild, but would be nice to know if the migration can handle that case.

calherries replied:
Isn't that case handled by lines 1013-1015?https://github.com/metabase/metabase/pull/28742/files/c9c911bace9294a62e63d816fce42a35adae091d#diff-41093075df90623837726c733afc60b57899a66d6e9878b4250413b74f045f67R1013-R1015

qnkhuat replied:
ah nice, sorry I didn't see that.

qnkhuat commented on resources/migrations/000_migrations.yaml:
```diff
@@ -14379,8 +14434,6 @@ databaseChangeLog:
                     nullable: false
 
 
-# >>>>>>>>>> DO NOT ADD NEW MIGRATIONS BELOW THIS LINE! ADD THEM ABOVE <<<<<<<<<<
-
```
can we have these back?

qnkhuat commented on src/metabase/sync/sync_metadata/fields/sync_instances.clj:
```diff
@@ -66,6 +66,7 @@
           :coercion_strategy         (when effective-type coercion-strategy)
           :semantic_type             (common/semantic-type field)
           :parent_id                 parent-id
+          :json_unfolding    json-unfolding
```
```suggestion
          :json_unfolding        json-unfolding
```

Commit:
Add false default value for json_unfolding

Commit:
Fix json-unfolding nil case

Commit:
Other suggestions

Commit:
whitespace

qnkhuat commented on src/metabase/driver/sql_jdbc/sync/describe_table.clj:
```diff
@@ -402,20 +405,33 @@
+                                                ;; use the database default
+                                                (let [existing-field (existing-fields-by-name (:name field))]
+                                                  (or (:json_unfolding existing-field)
+                                                      (and (nil? (:json_unfolding existing-field))
```
if we make `json_unfodling` non-nullable, we don't have to deal with it here.

qnkhuat replied:
then this predicate will be only `(:json_unfolding existing-field)`

calherries replied:
:json_unfolding can still be nil if the existing field doesn't exist. I've updated it here to reflect that. https://github.com/metabase/metabase/commit/4c53a5b5d92055d3a77d35b9012bfa03bf361c28

qnkhuat replied:
ah nice. that sound like a gotcha tho, should we have tests for it?

calherries replied:
It's covered by the `json-unfolding-default-false-test` test for json unfolding, since this code path is hit twice in that test. Once in the first sync when the json_unfolding default is false for the database, and again in the second sync when the json_unfolding default is true.

qnkhuat commented on src/metabase/models/database.clj:
```diff
@@ -177,7 +177,13 @@
-      (u/prog1 (handle-secrets-changes database)
+      (u/prog1  (-> database
+                    (cond->
+                      ;; If the engine doesn't support nested field columns, `json_unfolding` must be nil
```
where do we need things to behave differently with `false` and `nil`?

calherries replied:
When it comes to details json_unfolding=nil is treated the same as json_unfolding=true. See the implementation for json-unfolding-default in this namespace.
https://github.com/metabase/metabase/blob/f2e4d8e50a7f34bbe8e18a32cb967fa257d63eff/src/metabase/models/database.clj#L354-L357

qnkhuat commented on test/metabase/driver/mysql_test.clj:
```diff
@@ -602,20 +602,6 @@
                   (mt/db)
                   (t2/select-one Table :db_id (mt/id) :name "bigint-and-bool-table")))))))))
 
-(deftest can-shut-off-json-unwrapping
```
could you explain why we're removing this?

calherries replied:
Because of this change. https://github.com/metabase/metabase/pull/28742/files/5cf304e7b6964cba0f2d1c725b3c1abbe3dbd9fb#r1126403635

Now the logic is handled by `json-unfolding-default`, and tested here 
https://github.com/metabase/metabase/blob/f2e4d8e50a7f34bbe8e18a32cb967fa257d63eff/test/metabase/models/database_test.clj#L318

Commit:
Tidy describe-nested-field-columns

Commit:
Update comment

Commit:
Remove unused clear-nested-fields!

Commit:
Remove unused return value

Commit:
Fix H2 migration to use base_type not database_type

Commit:
Always set json-unfolding during sync

Commit:
Fix test

Commit:
Fix test

Commit:
Fix test

Commit:
Add comment explaining nested-field-column support for MySQL

Commit:
Merge branch 'master' into unfold-json-by-field

Commit:
Fix tests

Commit:
Fix tests

Commit:
Fix test

Commit:
Fix test

Commit:
Default json_unfolding to false for new fields

Commit:
Fix merge

Commit:
Add json_unfolding to mock tables

Commit:
Don't capitalize prepositions

Commit:
Update setting description

Commit:
whitespace

Commit:
whitespace

Commit:
Fix fetch_metadata

Commit:
Fix fetch_metadata

Commit:
Fix tests

Commit:
Fix test

Commit:
Fix clj-kondo

Commit:
Remove postgres database-supports test

Commit:
Fix postgres test

Commit:
Fix postgres test

Commit:
Fix mysql migration

Commit:
Merge branch 'master' into unfold-json-by-field

Commit:
Fix clj-kondo

Commit:
Don't test mariadb

Commit:
Fix mysql test

Commit:
Fix mysql json-unfolding nil case

Commit:
Add comments to test

Commit:
Add upterm step to mariadb test

Commit:
Move upterm step before tests

Commit:
Fix postgresql migration

Commit:
move upterm step to test-driver action

Commit:
Remove upterm step from drivers.yml

Commit:
Merge branch 'master' into unfold-json-by-field

Commit:
Comment out everything else in test-driver

calherries committed:
Merge branch 'master' into unfold-json-by-field

Commit:
Fix mariadb migration

Commit:
Remove upterm action

Commit:
Whitespace

calherries commented on test/metabase/driver/postgres_test.clj:
```diff
@@ -347,14 +348,6 @@
-                               (driver/database-supports? :postgres :nested-field-columns {:details details}))
-      {}                      true
-      {:json-unfolding true}  true
-      {:json-unfolding false} false)))
```
I removed this test because of changes to `database-supports?`. `json-unfolding-default-test` now covers it.

qnkhuat commented on src/metabase/db/liquibase/h2.clj:
```diff
@@ -11,6 +11,46 @@
 (defn- upcase ^String [s]
   (some-> s u/upper-case-en))
 
+(defn- create-udfs!
```
not sure how but should we have tests for this?

calherries replied:
I just updated this function so it only handles JSONObject. Now `migrate-field-json-unfolding-type-test` has 100% good coverage of JSON_VALUE. Migrations don't change, their tests don't either, and neither should the test coverage.

qnkhuat replied:
can you add a comment here saying it's tested in that migration?

Commit:
Only handle JSONObjects with JSON_VALUE, not JSONArray

qnkhuat commented on src/metabase/models/database.clj:
```diff
@@ -343,3 +349,8 @@
   ["databases" name name])
+
+(defn json-unfolding-default
+  "Returns true if JSON fields should be unfolded by default for this database, and false otherwise."
```
it'd be nice to give a bit of context why `nil` is treated as `true` here.

calherries replied:
Added on line 356-358

qnkhuat commented on src/metabase/sync/sync_metadata/fields/sync_metadata.clj:
```diff
@@ -73,6 +75,12 @@
+                          (common/field-metadata-name-for-logging table metabase-field)
+                          old-json-unfolding
+                          new-json-unfolding))
+           {:database_type new-database-type})
```
```suggestion
           {:json_unfodling new-json-folding})
```

qnkhuat replied:
and don't forget to add a test for this in `metabase.sync.sync-metadata.fields.sync-metadata-test`

calherries replied:
Fixed! https://github.com/metabase/metabase/commit/d51430fdbccf27e1597f6ae5c003ca6cb403f5d7

qnkhuat commented on test/metabase/driver/postgres_test.clj:
```diff
@@ -496,53 +489,55 @@
+                        (:take 1)
+                        (first)
+                        ((juxt :base-type :semantic-type)))))
+            (sync-tables/sync-tables-and-database! database)
```
why do we need to call these sync now?

calherries replied:
See line 540. `sql-jdbc.sync/describe-nested-field-columns` now uses the ID of the table arg, not just the name of the table. The Table for `describe_json_table` needs to be persisted to have an ID.

Commit:
Fix fields.sync_metadata/update-field-metadata-if-needed!

Commit:
Add comment explaining `json-unfolding-default`

qnkhuat commented on src/metabase/driver/sql_jdbc/sync/describe_table.clj:
```diff
@@ -402,20 +404,33 @@
+                                                ;; otherwise if the field doesn't exist
+                                                ;; use the database default
+                                                (let [existing-field (existing-fields-by-name (:name field))]
+                                                  (or (:json_unfolding existing-field)
```
do we have tests for this? I see we have `describe-nested-field-columns-test` in `describe_table_test.clj` but didn't see we updating that test.

Commit:
Add comment for JSON_VALUE test

Commit:
Update json-unfolding-default-true-test to use fresh db

Commit:
Fix test descriptions

Commit:
Don't capitalize setting display name

Commit:
Fix update json_unfolding

Commit:
Fix unfolding json for only one JSON column, not all

Commit:
Add e2e test for JSON unfolding setting

Commit:
Fix test

Commit:
whitespace

Commit:
Merge branch 'master' into unfold-json-by-field

Commit:
Fix test

Commit:
Fix json-unfolding for nested field columns

Commit:
Fix mysql and postgres tests for nested field columns

Commit:
Coerce nil json-unfolding to false for sync_metadata

Commit:
Add test for nil json-unfolding for sync_metadata

Commit:
Fix test

qnkhuat reviewed (approved):
I think the BE part looks good and the feature work as expected.

Commit:
Don't update json_unfolding from intial sync

Commit:
Add json_unfolding to update-field-test

calherries commented:
@qnkhuat I just fixed a bug in this commit https://github.com/metabase/metabase/pull/28742/commits/e0391951b4f1cb5e45662a0bbebac54aea0ed102. The problem was [these tests](https://github.com/metabase/metabase/pull/28742/commits/e0391951b4f1cb5e45662a0bbebac54aea0ed102#diff-3e624c65f922d3d133811640266aedea0c8666fb88c86661b0ce13f2ea7b89e9R757-R794) were failing because the json_unfolding value set by the user would be overwritten by a db sync. I've updated it so now only the initial sync will set the json_unfolding value, and json_unfolding won't be updated on subsequent syncs.

qnkhuat reviewed (approved):
Thanks for pinging, the new changes make sense to me. that's an easy case to overlook.

Commit:
Fix tests

kulyk commented on e2e/test/scenarios/admin/datamodel/field.cy.spec.js:
```diff
@@ -133,3 +135,52 @@ describe.skip("scenarios > admin > datamodel > field", () => {
+    );
+    // Go to field settings
+    cy.visit(`/admin/datamodel/database/${WRITABLE_DB_ID}`);
+    cy.contains("Many Data Types").click();
```
That should better be `cy.findByText("Many Data Types").click()`. If you need a substring, you can pass a regexp to `findByText`: `cy.findByText(/Many Data Types/i).click()`

kulyk commented on e2e/test/scenarios/admin/datamodel/field.cy.spec.js:
```diff
@@ -133,3 +135,52 @@ describe.skip("scenarios > admin > datamodel > field", () => {
+
+    // Check json is unfolded initially
+    cy.contains("json.a").should("be.visible");
+    cy.get("[data-testid='column-json'] [aria-label='gear icon']").click();
```
That should better be something like: 

```js
cy.getByTestId("column-json").within(() => {
  cy.icon("gear").click();
});
```

kulyk reviewed (approved):
Looks good! Could you please make the E2E test use [cypress-testing-library](https://testing-library.com/docs/cypress-testing-library/intro/) commands?

Commit:
Use cypress-testing-library commands

Commit:
Fix e2e test

Commit:
Fix e2e test
